File: backend/app/models/pipeline.py
"""
Pipeline model — defines a team workflow graph.
MongoDB collection: 'pipelines'

A pipeline is a directed graph where:
  - nodes  → teams (entry points or processing steps)
  - edges  → "when approved in team A, advance to team B"

When a node has multiple outgoing edges, the approver picks which
branch to follow (conditional routing — human-in-the-loop decision).
"""
from typing import Optional
from pydantic import BaseModel


class NodePosition(BaseModel):
    x: float = 0.0
    y: float = 0.0


class PipelineNode(BaseModel):
    id: str                      # React-Flow node id (e.g. "n-abc123")
    team_id: str                 # MongoDB team _id
    position: NodePosition = NodePosition()


class PipelineEdge(BaseModel):
    id: str                      # React-Flow edge id (e.g. "e-abc123")
    source: str                  # source node id
    target: str                  # target node id
    label: Optional[str] = ""   # shown to approver when branching


class CreatePipelineRequest(BaseModel):
    name: str
    nodes: list[PipelineNode] = []
    edges: list[PipelineEdge] = []


class UpdatePipelineRequest(BaseModel):
    name: Optional[str] = None
    nodes: Optional[list[PipelineNode]] = None
    edges: Optional[list[PipelineEdge]] = None

File: frontend/types/pipeline.ts
export interface PipelineNodePosition {
  x: number;
  y: number;
}

export interface PipelineNode {
  id: string;        // React-Flow node id
  team_id: string;
  position: PipelineNodePosition;
}

export interface PipelineEdge {
  id: string;        // React-Flow edge id
  source: string;    // source node id
  target: string;    // target node id
  label?: string;    // shown when multiple outgoing edges (conditional branch label)
}

export interface Pipeline {
  id: string;
  name: string;
  nodes: PipelineNode[];
  edges: PipelineEdge[];
  created_by: string;
  created_at: string;
  updated_at: string;
}

/** Branch option returned when an approved task has multiple outgoing edges. */
export interface PipelineBranchOption {
  edge_id: string;
  target_node_id: string;
  target_team_id: string;
  target_team_name: string;
  edge_label: string;
}
